src/game_player.cpp

```diff
@@ -712,17 +712,65 @@ bool Game_Player::CanWalk(int x, int y) {
 void Game_Player::BeginMove() {
 	int terrain_id = Game_Map::GetTerrainTag(GetX(), GetY());
 	const RPG::Terrain* terrain = ReaderUtil::GetElement(Data::terrains, terrain_id);
+	bool red_flash = false;
+
+	// Step Damage for States:
+	for (auto hero : Main_Data::game_party->GetActors()) {
+		for (auto state_id : hero->GetInflictedStates()) {
+			RPG::State *state = ReaderUtil::GetElement(Data::states, state_id);
+			if (state_steps_hp.size() < state_id) {
+				state_steps_hp.resize(state_id);
+			}
+			if (state_steps_sp.size() < state_id) {
+				state_steps_sp.resize(state_id);
+			}
+			if (state->hp_change_map_val > 0) {
+				if ((++state_steps_hp[state_id - 1]) >= state->hp_change_map_steps) {
+					state_steps_hp[state_id - 1] -= state->hp_change_map_steps;
+					hero->ChangeHp(-std::max<int>(0, std::min<int>(state->hp_change_map_val, hero->GetHp() - 1)));
+					red_flash = true;
+				}
+			}
+			if (state->sp_change_map_val > 0) {
+				if ((++state_steps_sp[state_id - 1]) >= state->sp_change_map_steps) {
+					state_steps_sp[state_id - 1] -= state->sp_change_map_steps;
+					hero->ChangeSp(-state->sp_change_map_val);
+					red_flash = true;
+				}
+			}
+		}
+	}
+
+	//Terrain:
 	if (terrain) {
 		if (!terrain->on_damage_se || (terrain->on_damage_se && (terrain->damage > 0))) {
 			Game_System::SePlay(terrain->footstep);
 		}
 		if (terrain->damage > 0) {
-			Main_Data::game_screen->FlashOnce(31, 10, 10, 19, 1);
-			Main_Data::game_party->ApplyDamage(terrain->damage, false);
+			for (auto hero : Main_Data::game_party->GetActors()) {
+				bool receive_damage = true;
+				for (auto object_id : hero->GetWholeEquipment()) {
```

Comment:
this complecated loop could be moved in a Game_Actor::IsImmuneToTerrainDamage function